### ISSUE TITLE
How to enter Service / Field Test Mode

### ISSUE THREAD

**Issue by E3V3A** [CellularPrivacy/Android-IMSI-Catcher-Detector#145]
This is not a real App issue, but may be helpful to understand how _Service Mode_ (**SM**) or _Field Test Mode_ (**FTM**) that can be reached on different devices, and on different firmwares (FW aka ROMs). 

The _SM_ is device **and** FW dependent, so what works on stock, usually don't work on a custom ROM, or completely different. The FW provide an OEM Java wrapper for this access, and AOSP, AOKP usually do not provide this wrapper, simply because they want it to work on all phones. 

The _SM_ is where you have direct modem access to all the internal and mobile network settings, that are not available in the AOS, API or UI. You can update FW, erase NV items, see all possible network parameters (that's what we want), factory reset phone, disable a multitude of other things, including bricking your phone! A typical (Samsung) SM screenshot may look something like this:

![0011s](https://cloud.githubusercontent.com/assets/194392/4593514/85e57e36-5086-11e4-886a-eacc8327b143.jpg)

For example, if you own an **HTC** device running stock FW, you may try these:

```
*#*#197328640#*#*       – Service Mode (SM)
*#*#7262626#*#*         = Field-Test Mode (FTM)
*#*#3424#*#*            = HTC Function Test
*#2263#                 – Service Mode
##3424#                 – Diagnostic Mode
##33284#                – Debug Menu : (Debug Screen, Test Calls, RC Data, HDR/1X Selection, Voice Privacy, DTMF Set, Korea Mode Set)
```

For newer **Samsung** devices you may try `*#0011*#`, or the above.

If you have other devices, do search on [XDA](http://forum.xda-developers.com/showthread.php?t=2744850). You can also try to search for new codes inside your stock firmware [by yourself](http://forum.xda-developers.com/galaxy-s3/general/ref-secret-codes-hidden-features-t1687249). 

---

**Q: Why is this important?** 
A: Because AOS doesn't provide a reliable way to get detailed neighboring cell info, so we're using the _multi-ril-client_ via OEM_HOOK requests to scrape the SM menu for details. So if you can get to this menu on any non-Samsung device, we would very much like to know which one, and the code you used, so that we can support it in our App. 
##

**Comment by LeoChirkov** [comment]
Do you need just to start a hidden menu from your app or do some more complex task? You may use [my app] (https://github.com/LeoChirkov/Service-Menu), which supports Sony, LG (not all), Motorola and HTC devices and can be easily extended to support more vendors. I'll be glad to do that and share the results with you, guys.

**Comment by SecUpwN** [comment]
Thanks for offering your help, @LeoChirkov. We'd essentially need to be able to access the `Service Mode` with the hidden details for scraping (like mobile network settings not available in the AOS, API or UI) - but that for as many devices as possible. Would access to those details be possible with your help?

**Comment by banjaxbanjo** [comment]
@LeoChirkov String samsungCode = "0011";
Add that to your app Leo a lot of samsung phones use this code to access service menu.

**Comment by LeoChirkov** [comment]
I didn't dig so deep to find out if it's possible to get direct access to such features as changing network settings. Here what I think about possible solutions:
1) start service menu explicitly and change settings manually (cons: factory FW only, not user friendly; pros: no permissions required, easy to implement);
2) get direct access to network settings (cons: only rooted devices, app must be installed as system; pros: good user experience; undefined: work on devices with custom ROMs).

**Comment by SecUpwN** [comment]
>only rooted devices, app must be installed as system

@LeoChirkov, have you tested our app yet? Press button `AT Command Interface` on your rooted device..

**Comment by LeoChirkov** [comment]
My bad, didn't test yet. Have you already tried to do some reverse engineering with service menu packages from any manufacturer?

**Comment by banjaxbanjo** [comment]
@LeoChirkov @SecUpwN access to service mode programmatically is possible all you have to do  is:
```Java
		Intent intent = new Intent();
		intent.setClassName("com.sec.phone", "com.sec.phone.SecPhoneService");
```
Only problem is you can't do this after 4.1 I think or less. To get the hooks to send to and get data back just logcat -b radio and manually open service menu then you will see every hook used depending what menu you go into. this data is sent back to the binded service and you can extract data from there.

**Comment by agilob** [comment]
>Only problem is you can't do this after 4.1 I think or less.

AIMSICD requires 4.1 or newer. It is worth to support only 4.1 which is used by 10%?

**Comment by banjaxbanjo** [comment]
It wouldn't but the phones from the 10% would have a pretty good advantage by having to access to phone service. Sucks really because we could get a crap load of data from this menu like timing advance,arfcn, tmsi, ciphermode (on certain phones).

**Comment by agilob** [comment]
Yes, but AIMSICD requires >= 4.1 and that snipped works only on <= 4.1 and selected phones with extra support from manufacturers (also some HTCs, LGs and Sonys). So the question is, is it important to support decaying version of android - 4.1 and some unspecified phones where this support can be added or removed with any update?

I would say definitely yes, but we would need a bigger group of testers with dozens different mobile phones to test it independently or spy on users and send their device details to remove server and analyse it.

First option is possible, but difficult, srlabs didn't mange to create a comprehensive list of devices, so it will be even harder for non-profit; second option: are we really considering it. Nope.

**Comment by banjaxbanjo** [comment]
I agree that is why this option is not coded into AIMSICD and has been known for some time that this was possible.

**Comment by banjaxbanjo** [comment]
checkout darshak and you will see how he used this method to dump ramdump files to extract gsm data.
